### PR TITLE
Fixing User Sets

### PR DESCRIPTION
**Pull request by drdrew42**
### General Changes to ResultSets
- `DB::ValidParams` and `DB::ValidDates` have been merged into a single `DB::Validate`
  - at the moment, it's branchy 🤮
  - I fully expect that to improve as other classes (besides UserSets) begin to utilize
- row creation calls need to be followed by `discard_changes` in order to pull in defaults from the DB layer

### Changes to User Sets
There are several major changes to how we handle user_sets:
- remove `type` column in the user_sets table
  - this is now determined by the problem_set relationship
  - 'UserSet' results and resultsets are still blessed into their respective subclass when retrieved
- default version is now `0`
  - meaningless for most problem set subclasses
  - for gateways (or other intentionally versioned sets) 0 will mean 'assigned', allowing 'attempted' versions to start from `1`
- user_problems are now created/deleted when a user_set is created/deleted
  - TBD: do we remove the API endpoint for creating user_problems?
  - TODO: adding (or removing) a problem to a problem_set should also automatically create (or delete) corresponding user_problems based on existing user_sets

### Changes to ProblemSets
- ProblemSets now have a method `validateOverrides` that handles:
  - dirty-ing up the ProblemSet columns, 
  - validating the set with the new overrides, 
  - discarding the changes. 
  - This subroutine also 'cleans' the override hash (in-place) of any key-values that are not actual overrides.

### Minor changes that should be extended
- in this PR, user_set -> problem_sets relationship name changed to singular: user_set -> problem_set
  - db relationship names should be _singular_ for `belongs_to` or `might_have` relations
  - db relationship names should be _plural_ for `has_many` or `many_to_many` relations
  - revisit all db classes to make these changes for consistency
- validation of overrides should happen with an existing Schema::Result (not `->new()`)
  - modify columns without database write, then check validation, finally `discard_changes`
- filter incoming parameters with list of relevant keys, rather than trying to list all possible irrelevant keys
- qw// with no left or right spacing
- no more `croak`ing
- signatures _everywhere_

### Regarding tests
- managing fields from csv files is abysmal
- coordinating undefined hash-keys vs. non-existent hash-keys is also abysmal
- we *really* need to comb through and ensure `is_deeply` uses the correct comparison order (_received_, _expected_)

### ~~Un~~finished tasks

- [x] db tests
- [x] mojolicious tests
- [x] test-stores

### CODE REVIEW

**Comment by drdrew42** [comment]
@pstaabp I broke more tests ;P

**Comment by drdrew42** [comment]
we'll also need to rebase onto #95 once it's merged, now that it's got some divergent commits.

File: lib/DB/Schema/Result/UserSet/HWSet.pm [hunk 1]
```diff
@@ -1,31 +1,23 @@
...
-	return DB::Schema::Result::ProblemSet::HWSet::valid_dates();
-}
-
-sub required_dates ($=) {
-	return DB::Schema::Result::ProblemSet::HWSet::required_dates();
-}
-
-sub optional_fields_in_dates ($=) {
-	return DB::Schema::Result::ProblemSet::HWSet::optional_fields_in_dates();
+sub valid_fields ($self, %args) {
```
Comment (pstaabp) [inline lib/DB/Schema/Result/UserSet/HWSet.pm:11]: I don't think the valid_fields, required and additional_validation are needed in the subclasses because of the new way you create the subclasses. Testing with print statements inside these are never printed and everything passes if removed.

**Comment by drdrew42** [comment]
I am not entirely convinced that our db/mojolicious tests are accurate. We do quite a lot of munging and that raises obvious concerns.

I think the csv files were quite good for building an initial set of users/courses/sets to test the UI -- but I now think we are quickly outgrowing their usefulness for unit-testing.

File: lib/DB/Schema/Result/UserSet/HWSet.pm (hunk 1 above)
Comment (drdrew42) [reply lib/DB/Schema/Result/UserSet/HWSet.pm:11]: I do believe you're right -- the UserSet subclasses can essentially disappear at this point. We may even be able to remove the `inflate_result` method as well... (that was only there because of the use of `->new`)

**Comment by pstaabp** [comment]
> We do quite a lot of munging and that raises obvious concerns.

In what ways?  Is this related to the cleanUndef? or sometimes it appears that we need to add/remove fields to match results?  If so, I think we can more careful with the tests.

**Comment by pstaabp** [comment]
I had introduced the `%args` to all of the subs in the `Validation.pm` module because at some point I had a second argument and I prefer using this than passing arguments in in a particular order.  With some refactoring since then, I think the only argument now is the `field_name` and perhaps we could return to only having `field_name` as the argument.  I can patch this up if this sounds good.

**Comment by drdrew42** [comment]
> In what ways?

It's unclear, when reading the test code, which fields are being specified and which aren't. I don't mean explicitly the 'cleaning up' or pruning of '*_id's, but the munging of the results is inconsistent (sometimes using Utils, sometimes a direct deletion) and as a result it's harder to trust that we're being accurate.

File: lib/DB/Schema/Result/UserSet/HWSet.pm (hunk 1 above)
Comment (pstaabp) [reply lib/DB/Schema/Result/UserSet/HWSet.pm:11]: All of tests run fine without that.  I thought that was key to handling the subclasses of UserSets.  I have to think on this.

Comment (drdrew42) [reply lib/DB/Schema/Result/UserSet/HWSet.pm:11]: UserSet params are validated by merging them onto their *existing* related ProblemSet (subclassed), which knows how to validate itself. The return on that validation is the 'cleaned' params (only those that differ from the related ProblemSet). The UserSet is then created from that 'clean' hash of params.

Nowhere in the process do we need to subclass a UserSet. So, it's safe, I presume, to dump all UserSet subclasses and associated methods.

Comment (drdrew42) [reply lib/DB/Schema/Result/UserSet/HWSet.pm:11]: Essentially, whenever a UserSet is requested, we should be returning a ProblemSet 'decorated' with the UserSet. The same should be applied when requesting a UserProblem (return a 'decorated' SetProblem).

**Comment by drgrice1** [comment]
I am still working on reviewing this, but I wanted to go ahead and comment on some things that I noticed.

The `valid_params` method in `lib/DB/Schema/Result/UserProblem.pm` is removed (or replaced with `valid_fields`) in this pull request.  However, the `valid_params` method of this class is still called in `bin/import_ww2_db.pl` on line 464.

There seems to be other problems with that script though.  For example, `$DB::Schema::Result::CourseUser::VALID_PARAMS` does not exist, but is referenced on line 186 of that file.  That causes an error.  This pull request is not responsible for this though.  

@pstaabp:  Signatures are used incorrectly in `bin/import_ww2_db.pl` (and elsewhere).   A method with the signature of `($=)` means the method takes a single optional parameter and the method doesn't care what the value of the parameter is.  Situations where that should be used is for parameters of a derived class method that doesn't use a parameter where the parent class does, or the reverse situation where a parent class defines a method but does not use a parameter but some child classes do.  It should not be used for a method that expects no parameters ever like you did for all of the methods in that file.

So to summarize the above, the `bin/import_ww2_db.pl` needs work.

**Comment by drgrice1** [comment]
> It's unclear, when reading the test code, which fields are being specified and which aren't. I don't mean explicitly the 'cleaning up' or pruning of '*_id's, but the munging of the results is inconsistent (sometimes using Utils, sometimes a direct deletion) and as a result it's harder to trust that we're being accurate.
> 
> Also, it's hard to keep referring back to a .csv and try to parse out what should and shouldn't be defined.

I have also had these thoughts about the tests.

**Review by drgrice1 — COMMENTED** [review]
Generally this looks good.  There are some minor things that should be fixed though.  Also, some consistency issues we need to perhaps discuss.

File: lib/DB/Schema/Result/CourseUser.pm [hunk 2]
```diff
@@ -132,21 +134,27 @@ __PACKAGE__->add_columns(
...
 
-sub required_params {
+sub required ($self, $field_name) {
 	return {};
 }
 
+sub additional_validation ($self, $field_name) {
+	return 1;
+}
+
```
[the remark this answers is not in the document]
Comment (drdrew42) [reply lib/DB/Schema/Result/CourseUser.pm:157]: Done 👍 good call

File: lib/DB/Schema/Result/ProblemSet/HWSet.pm [hunk 3]
```diff
@@ -3,101 +3,83 @@ package DB::Schema::Result::ProblemSet::HWSet;
...
+=head2 C<valid_fields>
 
-subroutine that returns the array for the valid dates: C<['open', 'reduced_scoring', 'due' ,'answer']>
+subroutine that returns a hash of the validation for both set_dates and set_params.
 
 =cut
 
-sub valid_dates ($=) {
-	return [ 'open', 'reduced_scoring', 'due', 'answer' ];
+sub valid_fields ($self, $field_name) {
```
[the remark this answers is not in the document]
Comment (drdrew42) [reply lib/DB/Schema/Result/ProblemSet/HWSet.pm:26]: I had this thought, and didn't pull the trigger on it. Will do. 👍

File: lib/DB/Schema/Result/Attempt.pm [hunk 4]
```diff
@@ -1,11 +1,10 @@
 package DB::Schema::Result::Attempt;
 use DBIx::Class::Core;
-use DB::WithParams;
 
 use strict;
 use warnings;
 
-use base qw(DBIx::Class::Core DB::WithParams);
+use base qw/DBIx::Class::Core DB::Validation/;
```
[the remark this answers is not in the document]
Comment (drdrew42) [reply lib/DB/Schema/Result/Attempt.pm:7]: At the moment, it does not -- but it does have json columns, and we have provided `valid_params` for that.

I don't know if there are any validations to make here -- moreover, I don't know if this is even how/where student attempts will be stored. This package will likely see significant overhaul in the near future.

What are your thoughts on inheriting from `DB::Validation` for (this and other) tables that have not yet implemented validation checks yet? (The current decision was based on prior inclusion of `WithParams` or `WithDates`...)

File: lib/DB/Schema/Result/ProblemSet.pm [hunk 5]
```diff
@@ -158,4 +159,51 @@ sub set_type ($set) {
...
+require validation, stripping out any unchanged values from the proposed update and then validating
+the set works with the proposed overrides.
+
+=cut
+
+sub validateOverrides ($set, $updates) {
+	foreach my $field_name (qw/set_dates set_params/) {
+		$set->_stripUnchanged($updates, $field_name);
+		$set->set_inflated_column(
+			"$field_name" => updateAllFields($set->get_inflated_column($field_name), $updates->{$field_name}));
```
[the remark this answers is not in the document]
Comment (drdrew42) [reply lib/DB/Schema/Result/ProblemSet.pm:175]: I'm not sure why I thought it was necessary ¯\_(ツ)_/¯

File: lib/DB/Schema/ResultSet/Attempt.pm [hunk 6]
```diff
@@ -2,7 +2,7 @@ package DB::Schema::ResultSet::Attempt;
 use strict;
 use warnings;
 use feature 'signatures';
-no warnings qw(experimental::signatures);
+no warnings qw/experimental::signatures/;
```
[the remark this answers is not in the document]
Comment (drdrew42) [reply lib/DB/Schema/ResultSet/Attempt.pm:5]: I think it was more to do with the fact that so many were already slashes. I have no significant preference for either.

File: lib/DB/Schema/Result/UserProblem.pm [hunk 7]
```diff
@@ -47,44 +42,53 @@ __PACKAGE__->add_columns(
...
+		default_value => 1,
 	},
 	problem_params => {
 		data_type          => 'text',
 		size               => 1024,
-		is_nullable        => 0,
 		default_value      => '{}',
 		serializer_class   => 'JSON',
 		serializer_options => { utf8 => 1 }
 	}
```
[the remark this answers is not in the document]
Comment (drdrew42) [reply lib/DB/Schema/Result/UserProblem.pm:56]: Oops, I thought I got them all (in this file at least). Now we've touched almost all of the `Result::` classes, and I didn't consider extending the change. I also prefer dropping it.

File: t/db/005_hwsets.t [hunk 8]
```diff
@@ -161,7 +161,7 @@ throws_ok {
 
 # Try to get a problem set that is not in a given course
 throws_ok {
-	$problem_set_rs->getProblemSet(info => { course_name => 'Precalculus', set_id => 6 });
+	$problem_set_rs->getProblemSet(info => { course_name => 'Precalculus', set_id => 999 });
```
[the remark this answers is not in the document]
Comment (drdrew42) [reply t/db/005_hwsets.t:164]: 😈 

This is a duplicated test. The set should exist, and belong to a course other than Precalculus.

File: lib/DB/Schema/Result/Attempt.pm (hunk 4 above)
[the remark this answers is not in the document]
Comment (drgrice1) [reply lib/DB/Schema/Result/Attempt.pm:7]: I thought that may be the case.  Perhaps just add a FIXME (or TODO) comment in the file for now so we know why it is that way, and that it needs work.

File: lib/DB/Schema/Result/UserProblem.pm [hunk 9]
```diff
@@ -22,74 +21,77 @@ __PACKAGE__->add_columns(
...
-		size        => 16,
-		is_nullable => 0,
+		data_type => 'integer',
+		size      => 16,
 	},
 	status => {
 		data_type     => 'decimal',
 		size          => '10,5',
-		is_nullable   => 0,
 		default_value => '0.00000'
```
Comment (pstaabp) [inline lib/DB/Schema/Result/UserProblem.pm:41]: It seems like for all of the fields with default_value, we are now also setting `retrieve_on_insert => 1`.  Should that go here too?

**Review by pstaabp — APPROVED** [review]

**Comment by pstaabp** [comment]
I didn't run `npm run test-stores` and am seeing problems on `users.spec.ts`.  Looking into this.

File: lib/DB/Schema/Result/UserProblem.pm (hunk 9 above)
Comment (drdrew42) [reply lib/DB/Schema/Result/UserProblem.pm:41]: good catch, I missed one! (this means we can stop worrying about `->discard_changes`

**Comment by drgrice1** [comment]
I am also seeing 2 failures in one test suite when I run `npm run test-stores` with mariadb.  They are the tests to 'Update a user' and 'Delete a global user' in the 'CRUD functions for global users' tests of `users.spec.ts`.

**Comment by pstaabp** [comment]
There was a change to the `addGlobalUser` which is causing this failure.  

In the future, I have some ideas to change to the `t/db/` tests to catch problems like this.

**Comment by drdrew42** [comment]
I handled the addGlobalUser issue -- but the deletion error is a mystery to me. We're going to change the way deletions are handled, and deleteGlobalUser is not yet implemented in the UI, so it's a push.

**Comment by pstaabp** [comment]
Everything is looking good now.

> We're going to change the way deletions are handled, and deleteGlobalUser is not yet implemented in the UI, so it's a push.

I put on the list of things to do on the backend improve the deleteXXX and related tests.
